### PR TITLE
`RemoveUnusedImports` creating ambiguous imports

### LINKED ISSUE #5703: `RemoveUnusedImports` creating ambiguous imports when there's explicit imports accompanying wildcards
## What version of OpenRewrite are you using?

org.openrewrite:rewrite-core:8.56.1

## What is the smallest, simplest way to reproduce the problem?

When using RemoveUnusedImports and GoogleJavaFormat together with overlapping wildcard imports, you may end up with ambiguous imports, making the code no longer compile.

Here's a test case:

```java
  @Test
  public void wildcardImports() {
    JavaParser.Builder<?, ?> parser =
        JavaParser.fromJavaVersion()
            .styles(ClasspathLoader.loadStyle("org.openrewrite.java.GoogleJavaFormat"));

    rewriteRun(
        spec ->
            spec.recipe(new RemoveUnusedImports()).parser(parser).expectedCyclesThatMakeChanges(2),
        java(
            """
            package com.helloworld;

            import java.sql.*;
            import java.sql.Date;
            import java.util.*;

            public final class Main {
              public static void main(final String... args) {
                final Date date = new Date(123);
                final java.util.Date date2 = new java.util.Date(123);
              }
            }
            """,
            """
            package com.helloworld;

            import java.sql.Date;

            public final class Main {
              public static void main(final String... args) {
                final Date date = new Date(123);
                final java.util.Date date2 = new java.util.Date(123);
              }
            }
            """));
  }
```

Resulting in:

```
org.opentest4j.AssertionFailedError: [Unexpected result in "com/helloworld/Main.java":
diff --git a/com/helloworld/Main.java b/com/helloworld/Main.java
index e2bf8f5..4374a31 100644
--- a/com/helloworld/Main.java
+++ b/com/helloworld/Main.java
@@ -1,6 +1,7 @@
 package com.helloworld;

 import java.sql.Date;
+import java.util.Date;

 public final class Main {
]
expected:
  "package com.helloworld;

  import java.sql.Date;

  public final class Main {
    public static void main(final String... args) {
      final Date date = new Date(123);
      final java.util.Date date2 = new java.util.Date(123);
    }
  }"
 but was:
  "package com.helloworld;

  import java.sql.Date;
  import java.util.Date;

  public final class Main {
    public static void main(final String... args) {
      final Date date = new Date(123);
      final java.util.Date date2 = new java.util.Date(123);
    }
  }"
	at org.openrewrite.test.RewriteTest.assertContentEquals(RewriteTest.java:622)
	at org.openrewrite.test.RewriteTest.rewriteRun(RewriteTest.java:511)
	at org.openrewrite.test.RewriteTest.rewriteRun(RewriteTest.java:130)
```

### PR DESCRIPTION
**Pull request by knutwannheden**
`RemoveUnusedImports` creating ambiguous imports

## Problem
The `RemoveUnusedImports` recipe was incorrectly creating ambiguous imports by unfolding wildcard imports for
types that were only used in fully qualified form. For example, when `java.util.Date` was used as `new
  java.util.Date()`, the wildcard import `import java.util.*;` would be replaced with `import java.util.Date;`
instead of being removed entirely.

## Root Cause
The original implementation lacked proper detection of qualified vs unqualified type references. During
wildcard import unfolding, the recipe would add imports for all types found in the `typesInUse` collection
without checking whether those types were actually used in unqualified form or only in fully qualified form.

## Solution
Enhanced the `RemoveUnusedImports` recipe in
`/rewrite-java/src/main/java/org/openrewrite/java/RemoveUnusedImports.java` with:

### 1. Efficient Unqualified Reference Collection
- **Added `collectUnqualifiedTypeNames()` method**: Performs a single AST traversal to collect all fully
  qualified names of types used in unqualified form
- **Performance optimization**: Changed from O(n*m) complexity to O(m + n) by replacing multiple AST traversals
  with a single pass and fast set lookups
- **Smart qualified reference detection**: Uses `isPartOfQualifiedReference()` helper to walk up the AST cursor
  and detect when an identifier is part of a `J.FieldAccess` chain

### 2. Improved Wildcard Unfolding Logic
- **Line 246-247**: Modified wildcard unfolding to filter types using
  `unqualifiedTypeNames.contains(fqType.getFullyQualifiedName())`
- **Lines 258-262**: Enhanced logic to completely remove wildcard imports when no types are used unqualified
- **Lines 282-284, 317**: Added defensive checks to prevent IndexOutOfBoundsException

### 3. Edge Case Handling
- **Added test case**: `keepImportWhenBothQualifiedAndUnqualifiedUsageExists()` verifies that imports are
  preserved when a type is used both qualified and unqualified in the same file
- **Mixed usage support**: Correctly handles scenarios where `Date` and `java.util.Date` appear in the same
  file

## Key Changes

**RemoveUnusedImports.java:**
- Lines 87-88: Added upfront collection of unqualified type names
- Lines 473-528: New `collectUnqualifiedTypeNames()` method with AST traversal logic
- Lines 502-524: Helper method `isPartOfQualifiedReference()` for qualified reference detection
- Line 246: Replaced inefficient per-type checking with fast set lookup

**RemoveUnusedImportsTest.java:**
- Lines 2215-2230: Added edge case test for mixed qualified/unqualified usage

## Verification
- ✅ Original failing test `onlyFullyQualifiedUsageShouldRemoveWildcard()` now passes
- ✅ New edge case test `keepImportWhenBothQualifiedAndUnqualifiedUsageExists()` passes
- ✅ All 63 existing `RemoveUnusedImports` tests continue to pass with no regressions
- ✅ Significant performance improvement for files with multiple wildcard imports

## Test Cases Covered
1. **Fully qualified only**: `java.util.Date` usage → wildcard import removed entirely
2. **Unqualified only**: `Date` usage → wildcard import unfolded to specific import
3. **Mixed usage**: Both `Date` and `java.util.Date` → import preserved due to unqualified reference
4. **Conflicting names**: Multiple packages with same class name → correct import precedence maintained

- Fixes #5703

### CODE REVIEW

**Review by timtebeek — APPROVED** [review]
Nice to see some of these longer standing issues resolved! I've applied the recipe here as well, and seems like we can now safely enable it going forward. Great!
